core/node/node_framework/src/service/mod.rs

```diff
@@ -159,67 +163,128 @@ impl ZkStackService {
+        let mut tasks_names: Vec<_> = join_handles.iter().map(|task| task.id()).collect();
+
         // Run the tasks until one of them exits.
-        let (resolved, _, remaining) = self
+        let (resolved, resolved_idx, remaining) = self
             .runtime
             .block_on(futures::future::select_all(join_handles));
-        let result = match resolved {
-            Ok(Ok(())) => Ok(()),
-            Ok(Err(err)) => Err(err).context("Task failed"),
+        // Extract the result and report it to logs early, before waiting for any other task to shutdown.
+        // We will also collect the errors from the remaining tasks, hence a vector.
+        let mut errors = Vec::new();
+        let task_name = tasks_names.swap_remove(resolved_idx);
+        match resolved {
+            Ok(Ok(())) => {
+                tracing::info!("Task {task_name} finished");
+            }
+            Ok(Err(err)) => {
+                tracing::error!("Task {task_name} failed: {err}");
+                errors.push(TaskError::TaskFailed(task_name, err));
+            }
             Err(panic_err) => {
                 let panic_msg = try_extract_panic_message(panic_err);
-                Err(anyhow::format_err!(
-                    "One of the tasks panicked: {panic_msg}"
-                ))
+                tracing::error!("Task {task_name}: {panic_msg}");
+                errors.push(TaskError::TaskPanicked(task_name, panic_msg));
             }
         };
+        tracing::info!("One of the task has exited, shutting down the node");
 
+        // Collect names for remaining tasks for reporting purposes.
+        // We have to re-collect, becuase `select_all` does not guarantes the order of returned remaining futures.
+        let remaining_tasks_names: Vec<_> = remaining.iter().map(|task| task.id()).collect();
         let remaining_tasks_with_timeout: Vec<_> = remaining
             .into_iter()
             .map(|task| async { tokio::time::timeout(TASK_SHUTDOWN_TIMEOUT, task).await })
             .collect();
 
         // Send stop signal to remaining tasks and wait for them to finish.
-        // Given that we are shutting down, we do not really care about returned values.
         self.stop_sender.send(true).ok();
         let execution_results = self
             .runtime
             .block_on(futures::future::join_all(remaining_tasks_with_timeout));
-        let execution_timeouts_count = execution_results.iter().filter(|&r| r.is_err()).count();
-        if execution_timeouts_count > 0 {
-            tracing::warn!(
-                "{execution_timeouts_count} tasks didn't finish in {TASK_SHUTDOWN_TIMEOUT:?} and were dropped"
-            );
-        } else {
-            tracing::info!("Remaining tasks finished without reaching timeouts");
+
+        // Report the results of the remaining tasks.
+        for (name, result) in remaining_tasks_names
+            .into_iter()
+            .zip(execution_results.into_iter())
```

Comment:
Nit: `.into_iter()` is redundant; `zip()` [accepts](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.zip) `IntoIterator` argument.